### ISSUE TITLE
Mattermost users can remove themselves from private channels via API and never rejoin them.

### ISSUE THREAD

**Issue by numericillustration** [mattermost/mattermost-server#8619]
#### Summary
Mattermost users connecting with irc clients via matterircd can remove themselves from private channels and never rejoin them.

#### Steps to reproduce
Anyone create a Direct Message private channel in the native desktop or web client via the "DIRECT MESSAGES    +" method

Invite one or more persons to the channel (this can even be yourself).

close the native / web client.

Connect via irc and you will be automatically joined to these channels.

issue a /part on the private channel.

Be unable to rejoin this channel or ever see messages to it.

#### Expected behavior
Users closing private channels do not leave them and can rejoin them

#### Observed behavior (that appears unintentional) 
Users creating a private channel via gui and then issuing /part in that channel are no longer able to join said channel.  This affects group private messages that are implemented as private channels as well.

In investigaing this I polled group private channels via API from users reporting this issue and found that groups affected were indeed missing a member, inlcuding a toy private message channel a user created only adding them selves to it.  

```
curl -H "Content-Type: application/json" -H "Authorization: Bearer $(grep Token foo.txt | awk '{print $2}')" https://<OUR SERVER UrL>/api/v4/teams/<TEAM ID>/channels/name/<channel name> | json 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   337  100   337    0     0   1840      0 --:--:-- --:--:-- --:--:--  1831
{
  "id": "1638m85ehibubdiq59grnxb6zw",
  "create_at": 1511815553818,
  "update_at": 1511815553818,
  "delete_at": 0,
  "team_id": "<TEAM ID>",
  "type": "P",
  "display_name": "<channel name>",
  "name": "<name>",
  "header": "",
  "purpose": "",
  "last_post_at": 1511815581022,
  "total_msg_count": 1,
  "extra_update_at": 1511815581017,
  "creator_id": "<user id of creator>"
}

curl -H "Content-Type: application/json" -H "Authorization: Bearer $(grep Token foo.txt | awk '{print $2}')" https://<OUR SERVER UrL>/api/v4/channels/1638m85ehibubdiq59grnxb6zw | json 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100   337  100   337    0     0   1711      0 --:--:-- --:--:-- --:--:--  1710
{
  "id": "1638m85ehibubdiq59grnxb6zw",
  "create_at": 1511815553818,
  "update_at": 1511815553818,
  "delete_at": 0,
  "team_id": "<TEAM ID>",
  "type": "P",
  "display_name": "<channel name>",
  "name": "sungo",
  "header": "",
  "purpose": "",
  "last_post_at": 1511815581022,
  "total_msg_count": 1,
  "extra_update_at": 1511815581017,
  "creator_id": "<user id of creator>"
}
curl -H "Content-Type: application/json" -H "Authorization: Bearer $(grep Token foo.txt | awk '{print $2}')" https://<OUR SERVER UrL>/api/v4/channels/1638m85ehibubdiq59grnxb6zw/members | json 
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100     2  100     2    0     0     14      0 --:--:-- --:--:-- --:--:--    14
[]
```

once they are no longer a memeber of a direct message private channel that has messages going to it the mattermost log shows entries like:
```
[2018/04/11 18:15:56 UTC] [EROR] /api/v4/channels/<channel ID>/posts:Permissions code=403 rid=eskimwe5ct84tbztrebm3pnabe uid=<user id> ip=<ip address> You do not have the appropriate permissions [details: userId=<user id>, permission=read_channel]
```

As a remediation effort I attempted to  re-add a user to the channel via API but the message returned was `"message": "Can not add user to this channel type",` with a 400 status.

the users are unable to recreate this group receiving an error ` A channel with that URL was previously created [details: id=<foo>, pq: duplicate key value violates unique constraint "channels_name_teamid_key"]

This /part command can even happen accidentally if their client automatically does this on close.  E.g wee chat until V 1.1.1 in 2015 https://github.com/weechat/weechat/issues/294 ) and undoubtedly others behave the same way.

#### Possible fixes
unknown

**Comment by numericillustration** [comment]
@saturninoabril It looks like I'm seeing this trouble with both `"type": "P",` channels and `"type": "D",` though [the API docs only show O and P](https://api.mattermost.com/#tag/channels%2Fpaths%2F~1channels%2Fpost).  

I seem to not be able to add users to either Direct or Private channels.  The [api allows removal of users from private channels](https://api.mattermost.com/#tag/channels%2Fpaths%2F~1channels~1%7Bchannel_id%7D~1members~1%7Buser_id%7D%2Fdelete) and clearly its possible to remove users from channels of type Direct as well since I have both types in my system with a user missing.  In both cases they are IRC users and one of them I have confirmed was subsequent to the `/part` command from the client.

This is an interesting loophole that one can create a Direct message channel with themselves, then via the API remove themselves from the channel and never be able to get back into it.  I just did this and now with the member list returning empty post delete, I can appear to join the channel in both the web and desktop clients but none of what I type in one client appears in the other and errors in the server logs like:
```
We could not get the unread message count for the user and channel [details: sql: Scan error on column index 0: converting driver.Value type <nil> ("<nil>") to a int64: invalid syntax]

msg: Script error. row: 0 col: 0 stack: null url:  [details: ]
```

It would be handy if [the docs](https://api.mattermost.com/#tag/channels%2Fpaths%2F~1channels~1%7Bchannel_id%7D~1members%2Fpost) also stated which types of channels could have users added to them.

**Comment by saturninoabril** [comment]
> [though the API docs only show O and P](https://api.mattermost.com/#tag/channels%2Fpaths%2F%7E1channels%2Fpost)

That's correct. There's separate API for creating [direct](https://api.mattermost.com/#tag/channels%2Fpaths%2F~1channels~1direct%2Fpost) and [group](https://api.mattermost.com/#tag/channels%2Fpaths%2F~1channels~1%7Bchannel_id%7D~1members%2Fpost) channels

> I seem to not be able to add users to either Direct or Private channels

It's by design that you can't add users to direct and group channels. For private channel, user must have necessary permission to do that - like permission to manage private channel members.

> This is an interesting loophole that one can create a Direct message channel with themselves, then via the API remove themselves from the channel and never be able to get back into it.

Working fine on me. Could you give repro steps? Also how did you get the error you've posted above?

> It would be handy if the docs also stated which types of channels could have users added to them.

Agree. Will send PR for this, or if you like, feel free to submit one.

**Comment by amyblais** [comment]
Hi @numericillustration Following up on this - did Saturnino's suggestions help?

**Comment by amyblais** [comment]
Hi @numericillustration, we haven't received an update so we'll assume that the problem is fixed or is no longer valid. 

If you still experience the same problem, try upgrading to the [latest version](https://about.mattermost.com/download/). 

If the issue persists, reopen this issue with the relevant information and we'd be glad to help you where we can.

**Comment by madchap** [comment]
If there was a solution to that, I am actually interested... :) Thanks for reaching out if you have it! Cheers.

**Comment by lfbrock** [comment]
@madchap - did you hit a similar issue? We haven't been able to reproduce it

**Comment by madchap** [comment]
Hi @lfbrock,

It seems so. I am using matterircd (https://github.com/42wim/matterircd) to connect to our enterprise mattermost, and I've done a `/part` on at least a known person, and no way to get it back. We had to create a private channel with only the 2 of us as a workaround.

**Comment by krnlpanik** [comment]
I can confirm this behavior with matterircd. /part shows a "user left the channel." system message in the DM window of the WebUI. Alerts no longer work correctly, and username doesn't automatically populate in the Direct Messages list, but the WebUI will CAN still issue messages to the channel if opened manually. I have only been able to restore proper functionality by manually adding the bi-directional membership rows to the channelmembers table which in quite a cumbersome task. 

My suspicion is that the API should reject or silently drop the request for removing channel membership of a D channel and instead is allowing it.

**Comment by madchap** [comment]
cc @BenoitKnecht (last comment)

**Comment by CountMurphy** [comment]
Having the same issue. ran a /part with matterircd on a person and now I can no longer communicate with them privately. Using the web interface, it just hangs. Using the mobile app gives "Refreshing messages failled." According to our admin, I'm generating permission denied logs whenever I try to access/send private messages with that person

**Comment by jasonblais** [comment]
@42wim would you have insights on the above? Seems like running `/part` on matterircd may prevent you from opening a direct message channel with them again.

**Comment by CountMurphy** [comment]
Workaround fix: 
 looking at the database there are "channels" for user to user
 you are don't have a record stating you are a member of the 
                  YourUser<->PersonYouParted's channel

Manually creating a record fixes this.

**Comment by krnlpanik** [comment]
I'm no SQL wizard, but I used the following query to come up with a list of affected D channels. This only lists the case of a two user D channel being reduced to one member. Multi-member D channels will not be listed here.
```
SELECT a.id,channels.name,cm.userid, users.username FROM users, channelmembers cm, channels, (SELECT channels.id, count(users.username) FROM channels, channelmembers, users 
WHERE 
    channels.type = 'D' AND
    channels.id = channelmembers.channelid AND
    channelmembers.userid = users.id 
GROUP BY CHANNELS.ID) a
WHERE a.count=1 AND channels.id=a.id AND cm.channelid=a.id AND cm.userid = users.id
```

**Comment by jasonblais** [comment]
@saturninoabril Can you help try the following repro steps:

1. Close your desktop app and browser.
2. Use this API call to remove yourself from a DM channel with User1: https://api.mattermost.com/#tag/channels%2Fpaths%2F~1channels~1%7Bchannel_id%7D~1members~1%7Buser_id%7D%2Fdelete
3. Reopen the desktop app and observe that the DM channel you removed yourself in step 2 is hidden in the sidebar.
4. Click "More.." at the bottom of the direct messages list and try to open a DM channel with User1
 - observed: DM channel with User1 doesn't open

/cc @numericillustration

**Comment by saturninoabril** [comment]
@jasonblais @numericillustration I've reproduced the issue and filed a ticket for it https://mattermost.atlassian.net/browse/MM-11593

**Comment by ThiefMaster** [comment]
great! what about providing a db migration step that automatically fixes such broken channel membership?

**Comment by Ppjet6** [comment]
I'd be interested in this as well.

**Comment by immolate-d** [comment]
I can confirm this issue is still happening, it's happening sporadically to all 200 people in my unit. Our administrator kept telling people last year that disappearing DMs were a known issue. I just saw it today, here is what happened:

My phone chime notification goes off
I see a mattermost notification on the screen
I unlock the screen and open mattermost
No no new DM is there. Completely gone
I opened a bunch of new DMs with people I didn't have current conversations with. After about 50 tries I found the right person, and there the message was from 10 mins ago.
I'm using Android. Mattermost Server Version 5.27.0 (Build 5.27.0)
App Version: 1.39.0 (Build 3000341)
